### ISSUE TITLE
Add a portal to see currently open windows

### ISSUE THREAD

**Issue by johan-bjareholt** [flatpak/xdg-desktop-portal#304]
I am a maintainer of the ActivityWatch project and we are polling the currently focused windows name and title. This is not an issue under Xorg but on Wayland this is a problem as there is no common API between compositors. I have discussed this shortly with both a wlroots and Gnome developer and they both seem to agree that exposing such data would be best solved by adding a xdg-desktop-portal API for this.

Wlroots and KWin already have APIs for this (gnome-shell too but it's disabled by default) but they are all different, so a xdg-desktop-portal API would significantly simplify things.

Suggestions on properties for windows, methods and signals that would be good to have:

Window properties:
- app_id (string)
- title (string)
- focused (bool)

Signals:
- FocusedApplicationChanged
- RunningApplicationsChanged

Methods:
- GetRunningApplications (array of windows)

Links to prior discussions with Gnome and wlroots developers:
- https://gitlab.gnome.org/jadahl/gnome-shell/commit/a31e33d5813dfe7c77c1a3a58c9f107d5242b1cb#note_339241
- https://gitlab.gnome.org/GNOME/gnome-shell/issues/983
- https://www.reddit.com/r/linux/comments/as1dd0/we_are_the_sway_wlroots_developers_ask_us_anything/egskqq9/

**Comment by TingPing** [comment]
It sounds very niche. What other applications would use this portal?

**Comment by johan-bjareholt** [comment]
Discord is another example, but another use-case would be for panels which can support multiple compositors (maybe not a good use-case for flatpak, but a use-case non the less). There are also competitors to ActivityWatch which are not open source (most notably RescueTime) which would have the same use-case as us.

I guess it's kind of niche, but not extremely niche. Sooner or later some API like this would be needed for wayland and to me it seems like xdg-desktop-portal is the best fit.

**Comment by matthiasclasen** [comment]
This is not the kind of information we would normally want to leak into sandboxes, I think.

**Comment by johan-bjareholt** [comment]
@matthiasclasen Neither should screenshots by default as that contains the same information and more, but you need to explicitly allow permissions for that right?

**Comment by matthiasclasen** [comment]
A screenshot is an explicit operation. You take a screenshot of an individual window, present the result to the user, and ask him: "share this with "SpyApp deluxe" ?" 

What you are asking for is hard to present to the user in a meaningful form.

**Comment by johan-bjareholt** [comment]
Why wouldn't it be possible to do the same with this? Show a prompt containin "SpyApp is requesting access for information about all your running applications" and give the user the ability to allow or deny the request. By default this request should probably only be valid for the current session but it would also be nice with a checkbox to remember this for future sessions as well.

This is how it is handled in macOS and I wish it was the same on Linux. However, on macOS if you allow it once it will allow it until you go into the settings and revoke the permissions, only allowing it for the current session by default seems more sane IMO. But that's just a small detail.

**Comment by TingPing** [comment]
It can certainly be done. A goal of portals has been not to pester users with yes/no questions like that though, but rather be tied to specific actions that make logical sense to cancel like choosing a file.

**Comment by amosbatto** [comment]
With all the recent press about smartphone addiction and the academic studies on how too much screen time is impacting social relations and mental health, there is a growing need for people to be able to track their daily activity on mobile devices and limit their screen time. Android and iOS now have this ability and mobile Linux in upcoming devices like the Purism Librem 5, Pine64 PinePhone and Necuno Mobile will need it. 

Please consider adding this feature. I know that I will use it in my Librem 5, and I'm sure that others will as well. I don't mind having to give a app permission to access this information, the first time that I use an app. We are going to need authorization of permissions like this if Linux is ever going to become a mobile OS that can compete with Android and iOS.

**Comment by hadess** [comment]
> Please consider adding this feature. I know that I will use it in my Librem 5, and I'm sure that others will as well.

It would probably be implemented in the host side on the Librem so you wouldn't need any portals, so it wouldn't help.

There's also no host API available right now. Somebody should really start with that, the portal can come later.

**Comment by johan-bjareholt** [comment]
> It would probably be implemented in the host side on the Librem so you wouldn't need any portals, so it wouldn't help.

The issue is that they have such an incredible amount of work to do already, it will take years until this feature would become a priority for them and then after that it will take yet more time until it's implemented and shipped. A third-party app could more easily

> There's also no host API available right now. Somebody should really start with that, the portal can come later.

[Wlroots own protocols](https://github.com/swaywm/wlr-protocols/blob/master/unstable/wlr-foreign-toplevel-management-unstable-v1.xml) actually has parts of the features I'm requesting, I will start prototyping with that tonight and see how it goes. I'm proposing this because there are more compositors out there and I don't want to have to implement a client for every one of them.

**Comment by ssokolow** [comment]
Bear in mind that it's not only smartphones where this can become a problem. I need this sort of tooling on my desktop PC, so I won't be switching away from X11 until this need is met.

**Comment by johan-bjareholt** [comment]
Now GNOME Shell has merged a DBus API for this but for yet another use-case, to be able to use the dogtail tool to automatically test GUI applications.

https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/326/diffs

And phosh (the Librem 5 shell) has implemented the rootston wlr-foreign-toplevel-management protocol https://source.puri.sm/Librem5/phosh/commit/532cfaf085cd440c3f849e92da8c8d65681c2a9c

Would be nice if we could have a solution which worked on both.

**Comment by matthiasclasen** [comment]
I'm still of the opinion that we don't want sandboxed apps to get into managing foreign toplevels. That is fundamentally a priviledged operation

**Comment by ssokolow** [comment]
Well, the demand isn't going to go away.

If you don't provide *some* compromise, you're likely to see an analogue to how, because Google refuses to allow things like YouTube downloaders in their extension store, users are becoming desensitized to being walked through the process of enabling side-loading and installing extensions which haven't had *any* vetting from a trusted third party.

I'd probably implement such a thing by writing an un-sandboxed daemon which wraps all the disparate APIs offered by various compositors and exposes a consistent API which can be whitelisted in the sandbox manifest... who knows whether I'd get it right from a security standpoint, but I'll do it nonetheless if remaining on X11 becomes unviable before the capabilities I rely on are officially offered.

**Comment by TingPing** [comment]
Lets not compare `xdg-desktop-portal` to restricting itself to "safe" information to Google's store policies...

The `flatpak` sandbox will always be strict that is the point and the discussion for what holes to allow in needs more merit than "We want it".

**Comment by ssokolow** [comment]
My concern is that, if the sandbox is ill-fitted for what users want their applications to do, it will have an outsized effect on how much sandboxing is actually done.

...and I do agree that vague permissions are something to be avoided. I'm just not sure whether it has to be something that would qualify as "vague" in this instance.

Something like "Read the title of your currently active window and get notified of changes to it" seems like a pretty clear thing that would be both intuitively obvious if a time-tracking application asked for it and suitably ominous if anything else did.

**Comment by TingPing** [comment]
OK So I'm still at my question personally. We have one application interested in using this. Anything else to add to this list?:

- ActivityWatch
- ~~Discord~~ (Lets be honest, it is proprietary and they don't care about Linux, I have no reason to believe they would ever support this API)

**Comment by ssokolow** [comment]
I write tooling of my own (ad-hoc at the moment) which would use it in ways comparable to ActivityWatch.

That's why I said that, if it's not available by the time X11 becomes unfeasible to continue using, I'll hack together some kind of "`extra-portals` daemon" of my own design.

**Comment by TingPing** [comment]
The difference with screenshots is those are very user centric actions that just happen one time. This is a more technical detail that will grant unlimited access in the background to potentially sensitive data that the user cannot verify at all. For example what if they open the web browser and the window title becomes "Your Name private-email@foo.org login" or something and the user leaked data they never intended to.

**Comment by ssokolow** [comment]
> Maybe we could limit that information to the app-id of each window (which is not easy to know always) never the window title. 

That would render it effectively useless for my use case.

I (and, I'd assume, ActivityWatch) need to be able to tell the difference between, for example, YouTube and Google Docs (in Firefox) or a Master's Thesis and a fanfic-writing project (in tools like FocusWriter).

**Comment by TingPing** [comment]
Honestly I think I convinced myself to be against it. Window titles are sensitive and there simply isn't a way to let a user interject themselves in that process.

Showing a persistent notification or tray isn't supported everywhere and I think users would just be trained to ignore such a thing. Showing when a Camera is active works because that is always privacy invading where as this is only sometimes so and users wouldn't realize it.

**Comment by TingPing** [comment]
It is OK that not everything will be Flatpak'd. Plenty of system level components cannot be and I think activity monitoring applications may fall under that.

**Comment by ssokolow** [comment]
I'd still prefer to sandbox everything I can, so I'll probably go ahead with my idea to implement some kind of "unofficial portals" daemon.

**Comment by luongthanhlam** [comment]
I'm developing a Vietnamese input method for IBus. It uses WM_CLASS to
identify what the window is focused in, so you can assign it to your
favorite input mode (e.g. pre-edit, surrounding text, US input, etc.). My
users like this feature very much, but sadly it does not work on Wayland.
So please bring back the WM_CLASS from X11 or something like that to
Wayland.

>

**Comment by johan-bjareholt** [comment]
> ActivityWatch
> Discord (Lets be honest, it is proprietary and they don't care about Flatpak, I have no reason to believe they would ever support this API)

If we add both a "width" and "height" property I believe it would be possible to add support for [dogtail](https://gitlab.com/dogtail/dogtail) on wayland for more shells than just Gnome Shell (Gnome Shell [currently has a DBus API](https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/326/diffs) for this)

**Comment by TingPing** [comment]
@ssokolow There are many ways to sandbox things. For example you can ship a systemd service with sandbox options.

**Comment by TingPing** [comment]
@luongthanhlam I believe your discussion belongs somewhere else. Portals are about user initiated actions and not low level wm details.

**Comment by ssokolow** [comment]
I got the impression systemd sandboxing wasn't as well-suited to desktop applications.

Am I mistaken or are you proposing that the monitoring and GUI be separate processes sandboxed using different technologies?

**Comment by TingPing** [comment]
You could split your daemon into its own process that would be a sandboxed DBus service. Your application could then be a Flatpak with permission to talk to it.

**Comment by ssokolow** [comment]
That was essentially what I was envisioning for my idea of writing some kind of "extra/unofficial portals" daemon.

**Comment by DanielJoyce** [comment]
Fed up with Wacom-style-tablet support, I am working on a user-space driver like sc-controlle but for digitize tablets.

Basically treating it as a mass of event sources and making setup even from the cli simple. For brand new tablets, bugs/deficiencies in the drivers means gnome wacom settings fails to load my tablet template, so there is no easy way to set mappings, even though all the raw event sources appear to work just fine.

One thing sc-controller provides, and I'd like to have in my user space table tool ( and can do under X11 ) is change button mappings for my digitizer tablet on a per-application basis ( gnome wacom doesn't provide this ). To do that I need information about the active window.

Some apps don't allow multiple shortcuts for a given action, so it would be awesome to have my tablet system go "Oh, Krita, so I need tablet key 1 to emit FOO key press to zoom in. Oh, now we're using myPaint, I need to emit BAR key press for zoom"

There needs to be some way to get this information, for input methods, usability, and accessability tools. 


Tji

**Comment by TingPing** [comment]
@DanielJoyce Messing with input of other applications seems extremely out of scope for Flatpak/Wayland/etc.

**Comment by DanielJoyce** [comment]
This is more about getting the window name, creating a uinput device can be managed outside wayland. Per-application behavior for tablets requires knowing the application/window name.

**Comment by TingPing** [comment]
If your software has permissions to control devices like that it would likely be outside of flatpak already.

**Comment by jefferyto** [comment]
I use [KeePass](https://keepass.info/) with its [auto-type](https://keepass.info/help/base/autotype.html) feature: When I press a system-wide hotkey, KeePass takes the title of the focused window, finds the corresponding username/password and enters this information into the window.

Or at least it use to work. After upgrading to Ubuntu 21.04 and switching the Wayland, neither part (getting the window title, entering key input) works. (I use this primarily to log onto websites and there are plugins that integrate KeePass with my browser, but I prefer to keep the two separate out of security concerns.)

There is a workaround for entering key input using `/dev/uinput`, but the developers of KeePass are [waiting for a common API](https://keepass.info/help/kb/autotype_wayland.html#info) for getting the window title instead of implementing individual solutions for different compositors.

KeePassXC is also [experiencing the same limitations](https://github.com/keepassxreboot/keepassxc/issues/2973#issuecomment-482708552). Auto-type is a very popular feature; the [Bitwarden feature request for auto-type](https://community.bitwarden.com/t/auto-type-autofill-for-logging-into-other-desktop-apps/158) is their most-voted feature request for desktop (third most-voted overall).

I hope it is clear this is a legitimate use-case for an application to access window titles.

**Comment by ssokolow** [comment]
I've also since realized that it's kind of hypocritical to refuse this when I only just [**accidentally** learned](https://github.com/flathub/com.corsixth.corsixth/issues/1#issuecomment-843940086) that an Open/Save portal's grant of access can be made persistent via the Documents portal and I'm *still* trying to formulate a Google query which will tell me where that information gets stored and how I can overrule (with an `incrontab` hack if necessary) so that programs don't have the ability to perpetually retain access to paths I Open/Save'd once without so much as a warning or checkbox.

**Comment by devmattrick** [comment]
To add another application, I'd like to have my keybinds automatically change depending on the active window using [key-mapper](https://github.com/sezanzeb/key-mapper/discussions/20).

**Comment by TingPing** [comment]
@rbreaves This sounds quite off-topic. Accessibility services run outside of the sandbox and are privileged processes.

**Comment by pktiuk** [comment]
> OK So I'm still at my question personally. We have one application interested in using this. Anything else to add to this list?:
> 
>     * ActivityWatch
> 
>     * ~Discord~ (Lets be honest, it is proprietary and they don't care about Flatpak, I have no reason to believe they would ever support this API)

You can add [AntiMicroX](https://github.com/AntiMicroX) to this list. Getting information about current window is required for auto profile - (switching gamepad profile dependent on currently selected window)

**Comment by thomasaull** [comment]
I recently wrote myself a small script to automatically change the button mappings on a wacom pen when I‘m im Blender and back. I‘m using xorg currently, where this is rather easy to do and not beeing able to do this under wayland is a bummer :-(

**Comment by zocker-160** [comment]
I am the developer of [keyboard-center](https://github.com/zocker-160/keyboard-center) and I would very much need this. 

Users need to be able to create application specific profiles and button mappings, which I can only implement when I know what application is currently active.

**Comment by jefferyto** [comment]
Looping @DReichl (for KeePass), @droidmonkey and @phoerious (for KeePassXC) into the discussion.

**Comment by Silve2611** [comment]
I also need this feature. 
There are tons of scenarios where it is necessary. 
Any application trying to read the screen for example. There is a reason this is possible on windows and on macos. 
Macos solved it quite good by asking for permission. 

If applications needing title and appname will never be allowed on linux with wayland, it will not be real os. Since when did Linux become the system of limited possibilities?
Asking for permission is totally fine but not allowing it at all seems like a very strong philosophy for me. It should be up to the users to decide how much an application is allowed to see.

**Comment by rbreaves** [comment]
> Asking for permission is totally fine but not allowing it at all seems like a very strong philosophy for me. It should be up to the users to decide how much an application is allowed to see.

I think it’s more about laziness than philosophy lol.. whether they admit that or not... Just very narrow mindedness on goals w/o much real consideration of how people actually use their computers.. that’d just slow them down & their contributions 😂.

**Comment by phoerious** [comment]
Thanks for tagging me. I read broadly through the comments. It is indeed absolutely beyond me why this kind of accessibility API is still but an afterthought if anything.

It was mentioned earlier that macOS has something like this already and I would indeed suggest going in the same direction. The first time you use Auto-Type in KeePassXC on macOS, a system popup appears prompting the user to allow access to accessibility and screen recording features (the permissions are badly named, but it works).

Such a prompt on Wayland, when approved by the user, should at least provide access to listing windows on the current desktop and their stacking order and allow for sending keystrokes to other applications. I can imagine a host of other applications that would need elevated accessibility capabilities, from screen recording apps over hotkey automation to screen readers. Some would even need to read the full window contents, not just the titles. AFAICS, there are already a few "portals" for the most essential things like screen recording and Orca also seems to be getting its data via DBus somehow, but it looks all rather underdeveloped to me. The result is that people implement this kind of feature via udev rules or suid daemons, which nullifies any of the encapsulation provided by Wayland, Flatpak, and the like entirely.

**Comment by pktiuk** [comment]
This is the most demanded feature on this repo (I think number to :+1: reactions is a good measure)   
I think developers should decide what to do with it in a bit more **official** manner.

For now we know that this feature is important for many types of apps, but there are concerns regarding the way of implementing and ensuring safety.

**Comment by Anyeos** [comment]
What about a list of IDs as hash? Something that is an ID of the related Window. If you only get a list of IDs there are no leaked information and the purpose of apps that can auto set a profile (like AntiMicroX) can be implemented consistently. 

An algorithm that returns always the same ID for the same App/Window combination. And if more information is needed it can "query" that ID to get it.

**Comment by ssokolow** [comment]
I suppose a "Please type the name of the window and install a Firefox plugin to intentionally leak the active tab's name (blame the designers of Wayland)" dialog is better than something like "Please run your applications under XWayland (blame the designers of Wayland)" or "Please enable accessibility in order to leak your active window information to everything that wants to claim to be a screen reader" for a time-tracking tool.

Honestly, this feels like a "Requests `filesystem=home`/`filesystem=host` for lack of a sufficiently sophisticated File Chooser Portal API" situation.

**Comment by rbreaves** [comment]
> or "Please enable accessibility in order to leak your active window information to everything that wants to claim to be a screen reader" for a time-tracking tool.

A) I think we shouldn't treat our users as though they are idiots.
B) Security does matter - but if you make it too onerous then people will reject it and find less secure ways to get around it.
C) Accessibility ought to be there for users of all kinds and we shouldn't be excluding them due to some idealistic ideology of a few programmers in the name of security.


If you want to know though Apple has an interesting approach that has an issue when it comes to their accessibility API which is if I set an Applescript to trigger on certain actions via BetterTouchTool it will sometimes cause the OS to think the trigger came from the App that has focus instead of BetterTouchTool so it actually requires the user to then add even more apps to security exclusions... that is actually pretty idiotic in itself. In Apple's attempt to increase security they are not even properly tracking what app is originating a trigger themselves, not sure what the solution ought to be in that case - but my point is that if you are going to implement heightened security then do it in away that still respects the user.

**Comment by matthiasclasen** [comment]
> Also if we want people with handicaps to be able to interact and work with Linux applications at any time in the future while using Wayland then this type of functionality will likely need to be implemented for them as well. We can't be sandboxing everything to the extent that automation, OCR and key remapping solutions such as key-mapper, xkeynsail, keyd and others are rendered useless the moment someone switches from x11 to wayland. We can do better to support situations where this level of access is required for legitimate use.

My opinion is that accessibility tooling is system functionality - it should live on the host side, not on the flatpak side.

**Comment by ssokolow** [comment]
I think the argument is that sandboxed or not, portals are the most appropriate place to handle things where the user should be prompted for permission and, if there's no portal, then it's going to be either proprietary to each desktop or a universally available hole in Wayland's envisioned security model.

**Comment by hammerandtongs** [comment]
What would a pull request look like for this issue?

Which existing portal looks most like what "most" people want?

https://github.com/flatpak/xdg-desktop-portal/tree/main/src

My sense is that the people are arguing at length in this thread are the only ones that are motivated to do anything about this but many seem to expect busy people that don't care about this issue to make the pull request...

**Comment by johan-bjareholt** [comment]
>  but many seem to expect busy people that don't care about this issue to make the pull request...

No, rather I'm not interested in wasting my time if I know that the PR will be rejected...

**Comment by rbreaves** [comment]
> Which existing portal looks most like what "most" people want?
> 
> https://github.com/flatpak/xdg-desktop-portal/tree/main/src

Good question, and I am not sure, I know so little about wayland if I am honest the way it is structured in contrast to X11. But I do agree with @johan-bjareholt that the main developers or contributors behind wayland and this xdg-desktop-portal claiming that this level of interactivity is niche ( @TingPing) or would be leaking too much info out of sandboxes ( @matthiasclasen  ) without much, if any thought, about accessibility concerns for the handicapped and use cases where it may be critical to someones work flow, to have dynamic key remaps as a person switching between apps, is concerning to say the least.

I am not a handicapped person, but I have the same need of simplifying shortcuts and workflows that might actually be quite similar to that of a handicapped person that uses various methods of keyboard input. Niche or not, what is being proposed here is very important and would be a serious quality of life improvement for far more people than they are making it out to be. It will absolutely prevent the adoption of rate of handicapped users, and people that require the use of uncommon keyboard input devices that may require additional awareness of what the actively focused application happens to be.

As long as this portal is lacking I do not see any Distro that runs on Wayland being a friendly or usable work environment for anyone that has to use additional software to help them with key input and it be application aware. Many of those users will have to either stick with x11 for the foreseeable future or switch to Windows or macOS.

**Comment by johan-bjareholt** [comment]
> My opinion is that accessibility tooling is system functionality - it should live on the host side, not on the flatpak side.

@matthiasclasen All use-cases for this feature are not related to accessibility, so I don't think it's as clear-cut. In macOS though it's considered an accessibility API. Regardless, as specified in the top comment I created an issue in the gnome-shell gitlab where they pointed me here. I will ask them again, since it was a few years ago after all.

**Comment by ssokolow** [comment]
> I will ask them again, since it was a few years ago after all.

Bear in mind that I have... strong feelings about GNOME's UI design decisions, so I'd be more likely to maintain a patchset for KDE Plasma (once its Wayland support stabilizes enough for me to get my multi-month-long login sessions that I currently rely on `kwin --replace` and Xorg's stability for), despite my extreme distrust of my ability to write correct C++ than to use a solution that's specific to GNOME Shell, and software I develop/maintain will follow suit in any desktops it supports intentionally rather than by accident.

(As they slowly drift GTK away from being "the universal toolkit" that I'd use to develop "apps for my KDE that also feel native on all other desktops", my attitude toward supporting GNOME grows more in line with the [Weird architectures weren't supported to begin with](https://blog.yossarian.net/2021/02/28/Weird-architectures-werent-supported-to-begin-with) perspective on people complaining about how pyca/cryptography's shift to depending on Rust broke niche platforms that used to work.)

**Comment by jadahl** [comment]
In my opinion, the kind of API asked for here (practically spy on user activity in the background) should in one way or the other be supported via portals, but I don't think in a traditional portal sense makes sense due to the severity and transparentness of how it'd work. A typical "do you want to fix the app (yes / no)" is not enough here.

Perhaps a better alternative is to make the portal lack the ability to query the user to allow the feature, but instead require these kind of things to be explicitly configured in e.g. Settings. The flow would be similar to how some Android permissions work these days:

 * Some portal provides a way to open e.g. a "accessibility settings" or "privacy" view. Lets call it the "system settings" portal. The portal backend would launch e.g. some view in gnome-control-center (in the GNOME case)
 * Another portal "user activity" that exposes e.g. the currently focused application, but without any way to ask to be enabled
 * Add a way to let applications advertise certain "accessibility" or "privacy sensitive" features e.g. in the .desktop file
 * Have said Settings application view list these applications in the above mentioned view, where one can explicitly toggle things like "Allow tracking user activity (Yes / No)"
 * On the first run, the applications explains to the user that they will be presented with an "accessibility" / ... settings, and that they need to find the app, and turn on the  tracking ability
 * The application asks the "system settings" portal to show the accessibility view
 * If the user decided to go through the trouble of finding the relevant thing to toggle, the "user activity" portal suddenly starts working, otherwise it'll remain non-functioning

**Comment by Silve2611** [comment]
> In my opinion, the kind of API asked for here (practically spy on user activity in the background) should in one way or the other be supported via portals, but I don't think in a traditional portal sense makes sense due to the severity and transparentness of how it'd work. A typical "do you want to fix the app (yes / no)" is not enough here.
> 
> Perhaps a better alternative is to make the portal lack the ability to query the user to allow the feature, but instead require these kind of things to be explicitly configured in e.g. Settings. The flow would be similar to how some Android permissions work these days:
> 
> * Some portal provides a way to open e.g. a "accessibility settings" or "privacy" view. Lets call it the "system settings" portal. The portal backend would launch e.g. some view in gnome-control-center (in the GNOME case)
> * Another portal "user activity" that exposes e.g. the currently focused application, but without any way to ask to be enabled
> * Add a way to let applications advertise certain "accessibility" or "privacy sensitive" features e.g. in the .desktop file
> * Have said Settings application view list these applications in the above mentioned view, where one can explicitly toggle things like "Allow tracking user activity (Yes / No)"
> * On the first run, the applications explains to the user that they will be presented with an "accessibility" / ... settings, and that they need to find the app, and turn on the  tracking ability
> * The application asks the "system settings" portal to show the accessibility view
> * If the user decided to go through the trouble of finding the relevant thing to toggle, the "user activity" portal suddenly starts working, otherwise it'll remain non-functioning

I agree with your approach. Althoug people dislike apple this is what macOS does and I think this solution should be looked at. Windows on the other hand has no restriction at all, which makes it easy for a developer but complicted in sense of data security.

In general Linux should not prohibid apps from appearing because they dislike them. They are used for a reason. In the EU it is now mandatory to track working times and the law has already passend many countries and will be standard in a few years. Linux will only be able to offer a time tracking solution if people use X11 which is bullocks…

I get the sandboxed Idea but we are protecting Linux with cussions. I personally feel as if a am at day care again and someone is telling me what I need to code and what I am not allowed.

That beeing said they should keep in mind that there are people with disabilities that rely on screen readers etc. Ignoring there needs now, will mean the end of Linux in any school, unversity public organization for good.

Placing security over accessibility is highly risky and morally questionable.

**Comment by PAStheLoD** [comment]
> Placing security over accessibility is highly risky and morally questionable.

+1

Not to mention that security as a practice, theory, process or goal to be attained is based on ever finer-grained separation of wanted an unwanted activity. (And that's why naive approaches just pop up the scary box of "this app wants to tinker with the system, do you allow it? if yes, type your password" - which just trains users to type their password into every and all input boxes until whatever they desired happens or they give up.)

Allowing the privilege of getting the app_id & title of the currently active window from the compositor is fine grained. The user (or in case of the majority of users a distribution that offers this as a feature) can control it better, because they can understand it better. (Do I want the "time tracker" to know which window is active? Yes, sounds reasonable. Do I want the time tracker to access all files? No, no thank you.  ... contrast this with the old paradigm:   Do I want the time tracker to "do whatever root can do"? Huh, the root of what? But sure, if it gets the thing working!)

Without these controls users are faced with a much bigger decision of whether to use something wayland based or not. (Which keeps X more and more alive, which fragments development effort.)



## offtopic :/

> "In the EU it is now mandatory to track working times and the law has already passend many countries and will be standard in a few years. [...] Linux will only be able to offer a time tracking solution if people use X11 which is bullocks…"

Determining time worked is not the same as spying on employees desktop activity. (For implementation details of the directive see https://ec.europa.eu/social/main.jsp?catId=706&intPageId=5115&langId=en , for example the 14 May 2019 ruling that deals with measurement.)

There are endless ways to implement systems, more and less privacy intrusive, that can offer "time tracking". From a simple webpage with a start-stop button, to arbitrarily complex systems that try to automatically label slices of time as work or non-work (let's say based on network activity, or by looking at keyboard and pointer input patterns, blablabla).

**Comment by ssokolow** [comment]
...and, as someone with both ADHD and an autism spectrum disorder, and the executive dysfunction that results from that, having the computer keep track of how much time I'm spending on various things with sufficient granularity to be able to distinguish different websites within the same browser (eg. YouTube vs. Google Docs) is critical so I can maintain perspective.

It's very much an "assistive technology" for me.

**Comment by Silve2611** [comment]
> 

I can agree on your point with asking for rights. This is what apple does very well. It asks you specifically for the right to read something. It even goes further and asks for specific rights for an application if your read more than the title. This should be the goto approach. 

As for the topic of time-tracking. They exist and people choose to use them. In the EU it is not even allowed to monitor the data of your employees so just use one that takes this into account.

Linux should not decide what kind of apps are allowed to exists.

Also this kind of decision is a discrimination which should not be something an os does…

**Comment by FrnchFrgg** [comment]
First of all this is not Linux deciding things. This is policy from the gnome and XDG projects.

What I find annoying is that the current state of things is that no sane solution is really possible currently because gnome decided to first restrict access to their API to a single caller (`xdg-desktop-portal-gnome`) «for security reasons» without thinking through any possible replacement or permissions system.

Even `root` can no longer access that information *from a non-sandboxed application*, which is good from a privacy point of view but problematic in some other cases.

My use case is an equivalent of the Android app «Family Link» which can tell me how long my ten-year-old son used which application. Using `ps` equivalents is not going to cut it because my son often leaves applications open in the background, especially Ardour and Blender. And in Firefox I want to distinguish between accessing his high school digital workspace to validate finished homework, and playing on Lichess.

I could use a more drastic approach disabling every and all non-approved content (with `apparmor` and stateful firewall rules even), but currently the contract we agreed on is «trust-based usage, but timed and supervised by adults». My `home-assistant` tells us when his PC is turned on and we routinely physically look at his screen throughout but he is not our only child and we cannot stay 24/7 in his room.

It is important to have the possibility to detect shorter «non-approved» uses and then confront him with the trust issues that this represent (that is doing our work as parents).
I could do so before, and now I cannot after applying an upgrade which I thought would be benign.

Note that my use-case could be seen as not relevant because parental control should work without explicit consent asked from the user. But I think that this on the contrary is part of the contract «for now you accept that we can see that information» and if our son reverts the authorization I can detect it and send a notice anyway.

**Comment by johan-bjareholt** [comment]
> What I find annoying is that the current state of things is that no sane solution is really possible currently because gnome decided to first restrict access to their API to a single caller (xdg-desktop-portal-gnome) «for security reasons» without thinking through any possible replacement or permissions system.

This is not true, I was messing around with it earlier today and there are two ways to access the Introspection API.
The first thing you mention is exactly what you say, that you need to be the xdg-desktop-portal-gnome application to be allowed to talk with the interface.
The second however if you read the source code, is if the org.gnome.shell.introspect option is set to true (and it's set to false by default). Just go into dconf-editor and set it to true and it should work.

**Comment by rbreaves** [comment]
> I agree with your approach. Althoug people dislike apple this is what macOS does and I think this solution should be looked at. Windows on the other hand has no restriction at all, which makes it easy for a developer but complicted in sense of data security.

Hogwash. Companies that use Windows have perfect security. /s :p

**Comment by FrnchFrgg** [comment]
> This is not true, [...] The second however if you read the source code, is if the org.gnome.shell.introspect option is set to true [...]
> https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/386d25e6f8ce11549526ea3776eb34138fcb3774/js/misc/introspect.js#L137

Your link is to an older version of `gnome-shell`. Nowadays [the authorization is deferred](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/misc/introspect.js#L140)
to `DBusSenderChecker.checkInvocation()` [which you can find in `utils.js`](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/misc/util.js#L545) which accepts `org.freedesktop.impl.portal.desktop.gtk` and `org.freedesktop.impl.portal.desktop.gnome` only, OR lets the call pass through if `global.context.unsafe_mode` is true. But that variable, contrary to a dconf setting, is transient only and is not settable from outside of `gnome-shell` debugger.

**Comment by johan-bjareholt** [comment]
@FrnchFrgg Oh my bad, sorry!

**Comment by FrnchFrgg** [comment]
To be very clear: I do not think that crippling the host ecosystem capabilities is flatpak's responsibility. And it should not be this project responsibility to "fix" it for gnome, which sends people here to ask for a portal even for non-sandboxed uses that they broke and are supposed to be outside of your scope.

Devising and implementing such an API is very much wanted but, as we can see in this bug, is hard to get right. I think that a lot of the pressure and discontent here comes from the fact that this broke even outside of flatpak sandboxes. Such pressure will not help your contributors to take time and write a system that is secure both on a computer science point of view and from a social engineering point of view.

And it means having people wait on your ability to devote time to a worthy feature request, but not necessary of utmost priority, so that they get a fix for their now broken use case which may be at best tangentially related to flatpak and sandboxed apps.

**Comment by Silve2611** [comment]
> To be very clear: I do not think that crippling the host ecosystem capabilities is flatpak's responsibility. And it should not be this project responsibility to "fix" it for gnome, which sends people here to ask for a portal even for non-sandboxed uses that they broke and are supposed to be outside of your scope.
> 
> Devising and implementing such an API is very much wanted but, as we can see in this bug, is hard to get right. I think that a lot of the pressure and discontent here comes from the fact that this broke even outside of flatpak sandboxes. Such pressure will not help your contributors to take time and write a system that is secure both on a computer science point of view and from a social engineering point of view.
> 
> And it means having people wait on your ability to devote time to a worthy feature request, but not necessary of utmost priority, so that they get a fix for their now broken use case which may be at best tangentially related to flatpak and sandboxed apps.

I do not agree that it is hard to do it. It just requires the correct attitude. Mac has already done it for several years so guidelines exists.

At the moment the security approach makes no sense. It prohibids everything pontentially dangerous, like a parent not allowing their children to run as they could fall.

**Comment by jefferyto** [comment]
From [An X11 Apologist Tries Wayland](https://artemis.sh/2022/09/18/wayland-from-an-x-apologist.html) by @faithanalog:

> I remember talking at length with the developer of [Talon Voice](https://talonvoice.com/), a voice control/eyetracking tool that works quite well on linux, about the challenges of supporting wayland. The other big thing, aside from how to do input emulation, was whether it was possible to query the list of windows and active focus for context-specific voice commands. I have definitely seen software that does this at this point, since most app panels are their own pieces of software independent of the compositor. And, as you’d expect from them, they can focus windows too. So given that these two problems are solved, it’s now on my list to try and help Talon get Wayland support when I have the energy.

**Comment by Silve2611** [comment]
> From [An X11 Apologist Tries Wayland](https://artemis.sh/2022/09/18/wayland-from-an-x-apologist.html) by @faithanalog:
> 
> > I remember talking at length with the developer of [Talon Voice](https://talonvoice.com/), a voice control/eyetracking tool that works quite well on linux, about the challenges of supporting wayland. The other big thing, aside from how to do input emulation, was whether it was possible to query the list of windows and active focus for context-specific voice commands. I have definitely seen software that does this at this point, since most app panels are their own pieces of software independent of the compositor. And, as you’d expect from them, they can focus windows too. So given that these two problems are solved, it’s now on my list to try and help Talon get Wayland support when I have the energy.

Well how is it done. We have all been talking about  the fact that we want to get window information about a certain windows but cannot do it.
Does he have a specific setting? Has something changed? How can the window title be received?

**Comment by hyuri** [comment]
AutoKey is another app example that would require knowing which app/window is in focus when a global shortcut is triggered.

**Comment by faithanalog** [comment]
author of 

> > From [An X11 Apologist Tries Wayland](https://artemis.sh/2022/09/18/wayland-from-an-x-apologist.html) by @faithanalog:
> > > I remember talking at length with the developer of [Talon Voice](https://talonvoice.com/), a voice control/eyetracking tool that works quite well on linux, about the challenges of supporting wayland. The other big thing, aside from how to do input emulation, was whether it was possible to query the list of windows and active focus for context-specific voice commands. I have definitely seen software that does this at this point, since most app panels are their own pieces of software independent of the compositor. And, as you’d expect from them, they can focus windows too. So given that these two problems are solved, it’s now on my list to try and help Talon get Wayland support when I have the energy.
> 
> Well how is it done. We have all been talking about the fact that we want to get window information about a certain windows but cannot do it. Does he have a specific setting? Has something changed? How can the window title be received?

I looked into it in more depth after writing that blog post and what I found is the software I had in mind (rofi) is using a wlroots extension called "wlr foreign toplevel management" for both getting the list of windows and focusing windows.

https://wayland.app/protocols/wlr-foreign-toplevel-management-unstable-v1

https://github.com/lbonn/rofi/blob/wayland/source/modes/wayland-window.c

I tried to figure out if there are gnome/KDE equivalents but wasn't able to find anything (this would have been half a year ago now, I think). That said I'm not an expert in wayland protocols, so perhaps someone else knows of something.

**Comment by johan-bjareholt** [comment]
@faithanalog The original post I wrote 4 years ago links to a reddit thread about just wlr-foreign-toplevel-management.

Unfortunately no one else than sway/wlroots wants to use that protocol, because it allows any Wayland application to read app names and titles which is a privacy risk.

**Comment by Mikenux** [comment]
Hello!

About leaking specific private data, it may be enough, when the app requests access to window tracking, to warn:
- That some window titles may contain information that may be considered private;
- If the app has means to share this information (e.g. inter-process/app communication, network access).

Another way is to have this access if the application cannot communicate with other applications and cannot access the network unless access is granted. For this, maybe it is possible to save the data in specific files known by flatpak (specific filenames, file format?) and for which only these are shareable (with the same app for synchronization or parental control, over the network)?

**Comment by yuhldr** [comment]
App calls sensitive permissions, warnings, enough. 

I want convenience and freedom, but if it's really for privacy, why should I use Linux?

 Accessibility features can also be called up by developers on Mac

**Comment by Mikenux** [comment]
@jadahl 

Is your proposal that the app wanting to track windows tells the user that it has privacy sensitive features and then open the privacy view in gnome-control-center (automatically or manually with a button when the app tells the user about its features?), and finally that the user should click on the row of the app and enable "Track User Activity"?

If so, that's a bit too much, as users will likely find the app and enable the setting for it. It might just annoy them, especially since they just want to use the app.

Besides, what's the point of doing that? Prevent users from automatically clicking "Allow" on dialog with "Allow" and "Deny" responses?

**Comment by Mikenux** [comment]
I got the answer by re-reading one of your comments on the accessibility portal issue.

So, I think a new design is needed to avoid the user automatically clicking an Allow-type response, but without annoying them with multiple steps.

Going back to the privacy aspect, as I said before, it is important to warn about the potential sharing of private data via internet access or communication with other processes (i.e. tell the user if there is this potential sharing and also when the app does not have this sharing capability). There are two solutions for this:
1. Tell the user and show them again the UI that asks to allow tracking if the app goes from having no means of sharing to having them (i.e. managing the permission change).
2. Require these apps to have no means of sharing and use a private data sharing portal (i.e. private data sharing is explicit and the app cannot have any internet access nor any inter-process communication). Such a portal, which does not exist, will require storing data in known files (to indicate to the user what data will be shared). Such a portal is perhaps more interesting in the long term (if it may exist).

I think this is very important because:
- knowing if the app can potentially share private data is more important than knowing if it collects such data.
- users will want to use the app they have installed, so will probably accept their activity being tracked anyway as the app will not be (fully) functional without this acceptance.

**Comment by ssokolow** [comment]
Any security expert will tell you that you're setting up a "false sense of security" situation.

Exfiltrating data can be accomplished in all sorts of non-obvious ways. For example:

* Drop it onto the clipboard briefly and then restore the old contents, synchronized to a time interval some other process will be listening on. (Tools like JDownloader's clipboard monitor and various Win32 tools for hooking text-drawing calls in Japanese visual novels and feeding them into translation tools on-the-fly are examples of legitimate uses for un-prompted clipboard interaction.)
* Save it in a known location on-disk that some other process will monitor for updates
* If using X11 or an XWayland setup where multiple applications share the same XWayland instance, store it in X11 window properties another process can watch and read. (The threat posed by combining an arbitrary code execution vulnerability with access to X11 is why `socket=fallback-x11` exists for Wayland capable applications.)
* Write it into the metadata (i.e. application name, stream description, etc.) of a PulseAudio connection for something else to read out using the API for writing audio mixer GUIs. (One of the classic ways to persist client-side state before modern HTML5 APIs was taking advantage of how much you can store in the `window.name` property in web browsers.)
* Engage in a "confused deputy attack" and feed it to something else with a vulnerability or misconfiguration which is authorized to do what you aren't. (This is why access to D-Bus must be limited as far as possible without breaking an application.)
* Hide it in the metadata of the user's next print job to be retrieved by some external tool. (Or, if you really want to get CIA-level, do what every color laser printer already does with things like the printer's serial number and steganographically encode it in the printed pages... color laser printers print patterns of yellow dots encoding their serial number and various other parameters intended to help the FBI track you down if you try to print money.)
* Make use of any device you the program has access to via the current need to enable `device=all` for access to webcams and joysticks/gamepads.
* Wait until the system is idle in a pattern that suggests the user is AFK or asleep and then do anything an external helper can watch and cover your tracks on. Perhaps streaming accumulated data into desktop notifications, depending on how the notification host is implemented.
* Hide the data inside the parts of structured files that aren't displayed WYSIWYG. (eg. custom chunks in PNG files, custom metadata fields inside JPEGs, custom files inside Zip-based formats like ODT, OOXML, EPUB, etc., custom data blocks inside stream container formats used in things like MPEG-family formats, Ogg, Matroska, Quicktime, ASF (WMA/WMV), etc.)

...or, with everyone feeling so confident about their security, go the [xkcd 538](https://xkcd.com/538/)-ish route and show the reader a desirable feature that seems to be legitimate evidence that the security model is getting in their way and needs to be circumvented for non-malicious reasons. (Stuff like how I've already seen some applications like keypress visualizers  for screencasts going straight to "Either run this under X11 or grant non-root users access to your keyboard's `evdev` device node" for want of a Wayland-level API.)

*Forcing* a "this or that but not both" permissions situation on users and application developers is a bad idea.

**Comment by Mikenux** [comment]
... It is about telling the app sharing the data from it, I don't see why it is giving a false sense of security. The reference point is the app, not external processes to it. If there are external processes watching or taking this data, it isn't relevant from the app that it is sharing the data. The app can store the data in metadata or in structured files, but it must know the other processes to use them to retrieve this data, no? Other permissions can be took into account for sure: that will be only remembering the user if the app is sandboxed or not, and any phrasing can be improved.

Otherwise, some areas can be certainly improved (data transmission over D-Bus, PulseAudio connection, device access, etc.), but it is not like I said to take it as is: those are only options, which are both discussable to know what to do in details.

And, sharing the data over the internet is not excluded in option 2: That's starting with a sandboxed app, then allow it to share the data over the internet (the "private data **sharing** portal"). Using a portal already generally means not using permissive permissions (e.g. generally, using the file chooser portal means not using filesystem=home). If this connection is mandatory, something then must be done to tell it appropriately.

**Comment by ssokolow** [comment]
I'm referring to option 2 for two reasons:

First, it's not feasible to retrofit "Require these apps to have no means of sharing" because there are so *many* APIs the things have to interact with and bad guys only need to find one of them... and they don't need to convince security auditors... they just need a solution that the average user won't recognize as a path for data exfiltration.

Java wasn't even retrofitting to the degree this is and it *still* had a couple of decades of applet security whac-a-mole before Java applets were finally required.

Things like JavaScript runtimes and WebAssembly can pull it off because they design their APIs from scratch to be simple enough to be audited.

Second, requiring people to go through special portals to get the API is reminding me of what I said recently regarding the idea of an xdg-pip Wayland extension. If you make your solution too onerous and restrictive, nobody will use it.

It's already hard enough to get applications to switch away from legacy permissions to portals and, as I said, GNOME's vision of Wayland is already driving application developers to circumvent the security model entirely to deliver the features users want.

**Comment by Mikenux** [comment]
In option 2, I'm talking about a sharing portal, which implies that the application, to use it, has no other way to share the data. If the app has permissive permissions to share data, there is no point in having a portal to share data, because a portal is built to replace the permissive permissions (including those that can be used in deviant ways ).

Alternatively, there's option 1, where it's about informing the user about the potential leak of private data from the app and asking the user again to grant permission if the app comes with more permissive permissions with an update.

Having one, then the other, depending on how sandboxing evolves over time, is also an option.

**Comment by ssokolow** [comment]
And my point is that "the application, to use it, has no other way to share the data" is an untenable position to enforce unless the entire API surface of the sandbox has been designed around it, the way something like WASI has, and attempting to enforce it will just imply to users that it *can* be done in a reliable manner.

**Comment by johan-bjareholt** [comment]
There is a new wayland protocol in the staging section called ext-foreign-toplevel-list[1] which allows clients to get all windows as well as their appid+title. This has of now only been implemented in a draft commit for the cosmic DE[2], hopefully more will follow. It has a lot of similarities with wlr-foreign-toplevel-management[3], but is more limited as it only shows all windows, and to be able to see which window is focused there is yet another protocol called foreign-toplevel-state[4].

What is more convincing about these two protocols compared to wlr-foreign-toplevel-management is that they have an intention to get them into wayland-protocols. There are still two big drawbacks however that still makes it unlikely that we will be able to use these two protocols anytime soon. First is simply that most compositors probably won't implement them. Secondly is that these protocols will only be accesible from so called "priviliged" clients. Exactly how to make a client "priviliged" will depend on the compositor and that is yet another big discussion.

Regardless, I am happy to see that at least something is happening in the wayland ecosystem in regards to this functionality.

[1] https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/staging/ext-foreign-toplevel-list/ext-foreign-toplevel-list-v1.xml
[2] https://github.com/pop-os/cosmic-comp/pull/76
[3] https://github.com/swaywm/wlr-protocols/blob/master/unstable/wlr-foreign-toplevel-management-unstable-v1.xml
[4] https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/196

**Comment by ssokolow** [comment]
> Exactly how to make a client "priviliged" will depend on the compositor and that is yet another big discussion.
> 
> Regardless, I am happy to see that at least something is happening in the wayland ecosystem in regards to this functionality.

I'm especially happy to see that someone's finally moving on the concept of privileged clients. That was promised over a decade ago as how the original Wayland concept would allow things like display control panels to not have to be reinvented as an in-process part of every new compositor.

**Comment by phoerious** [comment]
> There are still two big drawbacks however that still makes it unlikely that we will be able to use these two protocols anytime soon. First is simply that most compositors probably won't implement them. Secondly is that these protocols will only be accesible from so called "priviliged" clients. Exactly how to make a client "priviliged" will depend on the compositor and that is yet another big discussion.

That's a pretty heavy limitation for such a fundamental feature.

**Comment by pktiuk** [comment]
@matthiasclasen   
Lack of this API really hurts the functionality of Wayland for many use cases.    
This is the most demanded feature in this repository, which shows, that there is a real demand behind this request.   
It can be done securely and with respect to privacy. There should be an acceptable middle ground.

**Comment by kittenvr** [comment]
nyrna 

> It sounds very niche. What other applications would use this portal?
> 
> I know Discord is a common complaint that it can't see other apps, but since its proprietary somebody would have to convince them to use this anyway.

**Comment by ssokolow** [comment]
> nyrna

Do you mean this? → https://nyrna.merritt.codes/

**Comment by Mikenux** [comment]
> Yes. Currently, most application control is done via simulating keyboard macros (using the remote-desktop portal, for instance). In the future, there will be a tighter integration but more likely the other way around: Kando offers an IPC interface and third-party applications can use this to open custom menus.

Be aware that’s better when we have an idea of the data that is used. For example, if we know that you want to trigger an action within another app, then we can re-ask the user for a write permission.

> > and your app constantly running in the background.
> 
> Yeah, but this does not seem to be a problem.

This might change in the future. An app would be active only when it has the focus. For running in the background, permission will be necessary. However, for now, no need to worry about that.

**Comment by MythicHH** [comment]
It'd be nice if this would be implemented. I am trying to use an app called Nyrna which suspends other applications similar to how a console would and it only works with XWayland because of the lack of this portal.

**Comment by numfin** [comment]
I guess 7 years of people asking this feature is not enough to convince developers.
Thank you for caring about our safety.

On a serious thought: why not just give opt-in configuration so users who prefer functionality over safety - can use this feature?
